File: WorldFeed/Server/src/WorldFeed.Common/Controllers/ITrendsController.cs
namespace WorldFeed.Core.Controllers
{
    using System.Threading.Tasks;

    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Parameters.TrendsClient;
    using WorldFeed.Common.Web;

    public interface ITrendsController
    {
        Task<ITwitterResult<IGetTrendsAtResult[]>> GetPlaceTrendsAtAsync(IGetTrendsAtParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<ITrendLocation[]>> GetTrendLocationsAsync(IGetTrendsLocationParameters parameters, ITwitterRequest request);

        Task<ITwitterResult<ITrendLocation[]>> GetTrendsLocationCloseToAsync(IGetTrendsLocationCloseToParameters parameters, ITwitterRequest request);
    }
}

File: Sharebook/Client/src/app/core/Core/Controllers/ITrendsController.ts
import {inject, Inject, InjectionToken} from "@angular/core";

import {ITwitterRequest} from "../../Public/Models/Interfaces/ITwitterRequest";
import {ITwitterResult} from "../Web/TwitterResult";
import {IGetTrendsAtParameters} from "../../Public/Parameters/TrendsClient/GetTrendsAtParameters";
import {IGetTrendsLocationParameters} from "../../Public/Parameters/TrendsClient/GetTrendsLocationParameters";
import {IGetTrendsLocationCloseToParameters} from "../../Public/Parameters/TrendsClient/GetTrendsLocationCloseToParameters";
import {IGetTrendsAtResult} from "../../Public/Models/Interfaces/IGetTrendsAtResult";
import {ITrendLocation} from "../../Public/Models/Interfaces/ITrendLocation";
import {TrendsController} from "../../../controllers/Trends/TrendsController";
import {TrendsQueryExecutor} from "../../../controllers/Trends/TrendsQueryExecutor";

export interface ITrendsController {
  getPlaceTrendsAtAsync(parameters: IGetTrendsAtParameters, request: ITwitterRequest): Promise<ITwitterResult<IGetTrendsAtResult[]>>;

  getTrendLocationsAsync(parameters: IGetTrendsLocationParameters, request: ITwitterRequest): Promise<ITwitterResult<ITrendLocation[]>>;

  getTrendsLocationCloseToAsync(parameters: IGetTrendsLocationCloseToParameters, request: ITwitterRequest): Promise<ITwitterResult<ITrendLocation[]>>;
}

export const ITrendsControllerToken = new InjectionToken<ITrendsController>('ITrendsController', {
  providedIn: 'root',
  factory: () => new TrendsController(inject(TrendsQueryExecutor)),
});
